new blog post about Rocket Validator (#2636)
elixirschool/elixirschool: _posts/2021-09-02-rocket-validator.md

---
author: Jaime Iniesta
author_link: https://github.com/jaimeiniesta
categories: general
date: 2021-09-02
layout: post
title: "Validating Accessibility and HTML with Rocket Validator"
excerpt: >
Learn how we're checking and monitoring the new Elixir School site to detect and fix accessibility and HTML issues using Rocket Validator.
---

## Intro

One of the core principles of Elixir School is to be an inclusive community. We want to help people from all over the world learn to code using Elixir. To facilitate that, we provide translations of the lessons to many languages, maintained by volunteers worldwide. By removing the language barrier, we're helping more people get into the Elixir community.

Apart from the language barrier, there are other barriers that we're aware of - [web accessibility barriers](https://www.w3.org/WAI/people-use-web/abilities-barriers/) are everywhere on the World Wide Web, making it harder for people to access content that is often taken for granted.

> Low vision may make a web page unreadable if the text has low contrast or the font size cannot be adapted easily. Unsighted people can use screen readers to hear the page content, but it won't work well if the content is poorly structured and can't be properly interpreted by the screen reader software. People with reduced mobility may prefer to navigate through the lessons using the keyboard instead of the mouse, but again this won't work if the document is not laid out correctly.

Reviewer: ```suggestion
Low vision may make a web page unreadable if the text has low contrast or the font size cannot be adapted easily. Unsighted people can use screen readers to hear the page content, but it won't work well if it is poorly structured and can't be properly interpreted by the screen reader software. People with reduced mobility may prefer to navigate through the lessons using the keyboard instead of the mouse, but this won't work if the document is not laid out correctly.
```

Just to minor suggestions to promote conciseness

> The good news is that we can make Elixir School accessible for everyone. And it's not a hard task, it just requires that we're aware of those barriers and that we follow the web standards to provide well-structured web pages.

Reviewer: ```suggestion
The good news is that we can make Elixir School accessible for everyone. It requires that we're aware of those barriers and follows the web standards to provide well-structured web pages.
```

I would suggest avoiding saying that things "are easy" or that they're not challenging, everyone learns at its own pace, and that's fine. See https://justsimply.dev/ for more examples.

## How to find accessibility and HTML issues

> The World Wide Web Consortium (W3C) is an international community that defines and develops [Web standards](https://www.w3.org/standards/), HTML being probably the most used -and abused- of those standards. At the end of 1997, the [W3C launched](https://www.webdesignmuseum.org/web-design-history/w3c-html-validator-1997) the [W3C HTML Validator](https://validator.w3.org/), a service to check any web page and find issues in its HTML structure. More than two decades later this service is still around and in good health and many tools have been built around its [open source validator](https://github.com/validator/validator).

Reviewer: ```suggestion
The World Wide Web Consortium (W3C) is an international community that defines and develops [Web standards](https://www.w3.org/standards/), HTML being probably the most used -and abused- of those standards. At the end of 1997, the [W3C launched](https://www.webdesignmuseum.org/web-design-history/w3c-html-validator-1997) the [W3C HTML Validator](https://validator.w3.org/), a service to check any web page and find issues in its HTML structure. More than two decades later, this service is still in good health, and many tools have been built around its [open source validator](https://github.com/validator/validator).
```

> HTML is the basis of web pages, but still a perfectly valid HTML page can have accessibility issues. For example, a web page may have a bad combination of colors that results in low contrast: a light color text over a light background may lead to unreadable text by people with low vision. If a video has no captions, deaf users won't understand what it's talking about. Sections of the document may be not accessible for users that don't employ a mouse to move through the document. Zooming and scaling may be disabled, making it impossible for the users to enlarge the web page to see it better. Many known issues are easily detected using automated accessibility checkers like [axe-core](https://github.com/dequelabs/axe-core).

Reviewer: ```suggestion
HTML is the basis of web pages, but a perfectly valid HTML page can still have accessibility issues. For example, a web page may have a bad combination of colors that results in low contrast: a light color text over a light background may lead to unreadable text by people with low vision. If a video has no captions, deaf users won't understand what it's talking about it. Sections of the document may not be accessible for users that don't employ a mouse to move through the document. Zooming and scaling may be disabled, making it impossible for the users to enlarge the web page to see it better. Many known issues are easily detected using automated accessibility checkers like [axe-core](https://github.com/dequelabs/axe-core).
```

## The power of automated site validation

> So, it's great that we have tools that help in detecting issues on web pages. Both the W3C HTML validator and axe-core are open source and can be incorporated into our web development flow. But, there's an important limitation: they can only check one URL at a time. When you have a small website with just a few web pages, it might be an option to do manual checks using these tools. But, what can we do with a large site like Elixir School? According to the latest reports, Elixir School has more than 1,000 web pages. If we want to check them all with the W3C Validator and axe-core, we would need to run 2,000 manual checks!

Reviewer: ```suggestion
It's great that we have tools that help in detecting issues on web pages. Both the W3C HTML validator and axe-core are open source and can be incorporated into our web development flow. But, there's an important limitation: they can only check one URL at a time. So, when you have a small website with just a few web pages, it might be an option to do manual checks using these tools. But, what can we do with a large site like Elixir School? According to the latest reports, Elixir School has more than 1,000 web pages. If we want to check them all with the W3C Validator and axe-core, we would need to run 2,000 manual checks!
```

## 🚀 Rocket Validator to the rescue!

> [Rocket Validator](https://rocketvalidator.com) is an online service that performs site-wide validation of large websites. It's basically an automated web crawler that will find the internal web pages of a site and validate each web page found using the W3C Validator and axe-core, producing detailed reports on the issues found, within seconds.

Reviewer: ```suggestion
[Rocket Validator](https://rocketvalidator.com) is an online service that performs site-wide validation of large websites. It's an automated web crawler that will find the internal web pages of a site and validate each web page found using the W3C Validator and axe-core, producing detailed reports on the issues found, within seconds.
```

> This service, also powered by Elixir, Phoenix, LiveView and Oban, is a subscription-based service for web agencies and freelance developers. But, it's free for open source projects like [Elixir School](https://github.com/elixirschool/school_house/pulls?q=is%3Apr+label%3A%22rocket+validator%22+is%3Aclosed) and [Docusaurus](https://github.com/facebook/docusaurus/pulls?q=is%3Apr+rocket+validator)! We're using Rocket Validator to run reports on the new Elixir School and to monitor the site via weekly-scheduled and deploy-triggered validations.

Reviewer: ```suggestion
This service, also powered by Elixir, Phoenix, LiveView, and Oban, is a subscription-based service for web agencies and freelance developers. But, it's free for open source projects like [Elixir School](https://github.com/elixirschool/school_house/pulls?q=is%3Apr+label%3A%22rocket+validator%22+is%3Aclosed) and [Docusaurus](https://github.com/facebook/docusaurus/pulls?q=is%3Apr+rocket+validator)! We're using Rocket Validator to run reports on the new Elixir School and monitor the site via weekly-scheduled and deploy-triggered validations.
```

## Running a site validation report

> To run a site-wide validation report, we just need to enter a starting URL. This is usually the front page of the site we're going to validate, but it can be any internal URL, or an XML or TXT sitemap containing the web pages to validate. Rocket Validator will visit this initial URL and discover the internal web pages, adding them to the report.

Reviewer: ```suggestion
To run a site-wide validation report, we need to enter a starting URL. This URI is usually the front page of the site we're going to validate, but it can be any internal URL, or an XML or TXT sitemap containing the web pages to validate. Rocket Validator will visit this initial URL and discover the internal web pages, adding them to the report.
```

![Rocket Validator new site validation report](https://www.dropbox.com/s/amt7ilhw3b0sdh2/rocket-validator-new-report.png?raw=1)

We can also define the validation speed as a rate limit of requests per second and define the options:

* **Check HTML**. Will run the W3C HTML Validator on each page found.
* **Check Accessibility**. Will run axe-core on each page found.
* **Deep Crawl**. Will find more web pages on the site by following recursively the internal links found.

Once we click on **Start Validation**, results will appear within seconds as the web pages are validated. We can browse the Summary Report to see a global overview including the most important issues (so you know what to fix first), as well as reports that group the common issues on the site and detailed reports per each web page.

![Rocket Validator summary report](https://www.dropbox.com/s/z611nr8ofpikawx/rocket-validator-summary-report.png?raw=1)

## Sharing reports

> Any site validation report generated by Rocket Validator can easily be shared, just by pasting its URL. This lets us reference any section of the report in Github issues, so we can work together on fixing them.

Reviewer: ```suggestion
Any site validation report generated by Rocket Validator can easily be shared just by pasting its URL. This lets us reference any section of the report in Github issues, so we can work together on fixing them.
```
